### ISSUE TITLE
Stalled job not requeued

### ISSUE THREAD

**Issue by spearmootz** [taskforcesh/bullmq#443]
I had a job that reported as "Stalled".  In redis there was a lock, an entry in the stalled key.  Nothing seemed to be wrong, but the stalled job did not reprocess and the subsequent jobs did not process either.

the worker is a sandboxed worker, the original work process did not break or have an error from what i can see.

I have a queue class that automatically instantiates queue, the worker shares a connection with the queue.   then i have a queue scheduler and queue events all with their own connection.

any ideas? 

not sure if this makes a difference but i do make use of job.waitUntilFinished(this.queueEvents)

**Comment by manast** [comment]
> I had a job that reported as "Stalled"

How was it reported as stalled?

**Comment by spearmootz** [comment]
<img width="779" alt="Screen Shot 2021-03-29 at 2 44 12 PM" src="https://user-images.githubusercontent.com/19737290/112897756-3b0e7300-909d-11eb-9fa5-6ed7f2db96e5.png">
<img width="571" alt="Screen Shot 2021-03-29 at 2 44 25 PM" src="https://user-images.githubusercontent.com/19737290/112897777-42ce1780-909d-11eb-9102-1beefc641e71.png">
<img width="512" alt="Screen Shot 2021-03-29 at 2 44 36 PM" src="https://user-images.githubusercontent.com/19737290/112897797-49f52580-909d-11eb-9f04-b70e3974ee06.png">
<img width="562" alt="Screen Shot 2021-03-29 at 2 44 57 PM" src="https://user-images.githubusercontent.com/19737290/112897842-57121480-909d-11eb-996b-ea3405e7792a.png">
<img width="349" alt="Screen Shot 2021-03-29 at 2 45 40 PM" src="https://user-images.githubusercontent.com/19737290/112897916-70b35c00-909d-11eb-80c9-00d383b3fb0e.png">

**Comment by spearmootz** [comment]
Let me know if there is anything else i can provide. love bullmq btw.

**Comment by manast** [comment]
are you sure your QueueScheduler is connected and working?

**Comment by spearmootz** [comment]
This is my class. should i be listening to a disconnect event or something along those lines?

if you see on instastiation i create a queueScheduler with its on connection.

```export default class HfsQueue<QueueData, QueueReturnValue> {
  static getWorker(name: string, processor: Processor | string, options: WorkerOptions = {}) {
    return new Worker(name, processor, {
      //@ts-ignore
      connection: getRedisConnection(),
      ...options,
    });
  }

  private _name: string;
  private _queueScheduler?: QueueScheduler;
  private _queueEvents?: QueueEvents;
  private _queueOptions?: QueueOptions;
  private _queue?: Queue;
  private _connection: Redis;

  constructor(name: string, queueOptions: QueueOptions = {}) {
    this._name = name;

    this._queueOptions = {
      ...queueOptions,
      defaultJobOptions: {
        ...defaultJobOptions,
        ...(queueOptions.defaultJobOptions || {}),
      },
    };

    this._connection = getRedisConnection();

    this._queueEvents = new QueueEvents(this._name, {
      connection: redisConnectionOptions,
    });

    this._queueScheduler = new QueueScheduler(this._name, {
      connection: redisConnectionOptions,
    });
  }

  get connection() {
    return this._connection;
  }

  get queueScheduler() {
    return this._queueScheduler;
  }
  get queueEvents() {
    return this._queueEvents;
  }

  get queue() {
    if (this._queue == null) {
      this._queue = new Queue(this._name, {
        //@ts-ignore
        connection: this.connection,
        ...this._queueOptions,
      });
    }

    return this._queue;
  }

  getWorker(
    processor: Processor | string,
    options: WorkerOptions = {},
  ): Worker<QueueData, QueueReturnValue> {
    return HfsQueue.getWorker(this._name, processor, {
      //@ts-ignore
      connection: this.connection,
      ...options,
    });
  }

  waitUntilJobFinished(job: Job) {
    return job.waitUntilFinished(this.queueEvents);
  }

  enqueue(name, data: QueueData) {
    return this.queue.add(name, data);
  }

  async enqueueAndGetJobAfterCompletion(name, data: QueueData) {
    const job = await this.enqueueAndWaitForCompletion(name, data);
    return this.getUpdatedJob(job);
  }

  async enqueueAndWaitForCompletion(name, data: QueueData) {
    const job = await this.enqueue(name, data);
    await this.waitUntilJobFinished(job);
    return job;
  }

  getUpdatedJob(job: Job): Promise<Job<QueueData, QueueReturnValue>> {
    return Job.fromId(this.queue, job.id);
  }

  async getUpdateJobAfterCompletion(job: Job) {
    await this.waitUntilJobFinished(job);
    return this.getUpdatedJob(job);
  }

  close() {
    if (this.queueScheduler != null) {
      this.queueScheduler.close();
    }

    if (this.queue != null) {
      this.queue.close();
    }
  }
}```

**Comment by manast** [comment]
difficult to know just from this code, I would attach some event listeners to _queueScheduler and its connection to see if it is behaving correctly.

**Comment by spearmootz** [comment]
any suggestions, trying to be independent here to find the source of the problem.

i tried to do something like getEventNames() and it returns an empty array. i tried to subscribe to all the keys in the queue-base with no luck.

Now i access the redis client and open a monitor and log everything but its a lot of data

**Comment by manast** [comment]
so can you attach listeners to 
```typescript
_queueScheduler.connection.client.on("connect", () => console.log("connected"))
```
listen to other events such as error, disconnect, etc. they are documented here: https://github.com/luin/ioredis#connection-events

**Comment by manast** [comment]
if everything is correct you should see the console logs of the events, if not, something is wrong in your connection, also make sure you are connecting to the correct redis instance, it is quite common to make the mistake of using the wrong instance...

**Comment by manast** [comment]
no, that is very strange, it does not make any sense that the connection created inside QueueScheduler is not connecting.

**Comment by manast** [comment]
I would put console logs inside the QueueScheduler code to see why the connection created inside is not correct. Have you attached more events besides connect? like error, etc?

**Comment by spearmootz** [comment]
ok, so i have done that and what i find is that the connection connects, there are no errors at any point. one thing i find interesting is that i only see that this.run() gets only called once.

**Comment by spearmootz** [comment]
so i intercept the redis connection you pass to queue base that then passes it to redis_connection and do the following

['connect', 'ready', 'error', 'close', 'reconnecting', 'end', 'wait'].forEach((event) => {
        newConnection.on(event, () => console.log(`queueScheduler EVENT: ${event}`));
      });
      
      i also put a console log when this.run is called and here is the resulting logs
      
queueScheduler EVENT: connect
queueScheduler EVENT: ready
run queue scheduler

**Comment by manast** [comment]
that seems to be correct

**Comment by manast** [comment]
Inside the QueueScheduler, this call is the one that will move stalled jobs back to wait:
```ts
      await this.moveStalledJobsToWait();
```

**Comment by spearmootz** [comment]
ok! i will go ahead and put the console log there and watch it in production.  ill reply back tomorrow with results!

**Comment by manast** [comment]
@mfrye let me know what you like me to add to this please https://docs.bullmq.io/guide/jobs/stalled

**Comment by mfrye** [comment]
@manast I'm not if you can answer it, but potentially low level details, or at least point me in the right direction where I can find them, into exactly how CPU could affect the stalling.

In my case, jobs are either failing or legitimately stalling, yet they are never removed from the active queue.

I'm up to date with the latest version, have followed the docs exactly around sandboxing, separate redis connections, etc, yet my queue is stalling and completely clogging up.

I want to help debug this, but I'm not sure where I should be looking... In the queue scheduler code, the worker, the redis scripts?

**Comment by mfrye** [comment]
I think I've narrowed down what's happening to somewhere in the processJob worker, related to the lockExtender function. Jobs fail or stall, but the lock timer isn't being cleared out. It just keeps going and extending the lock indefinitely. 

I'll try to see if I can figure out a fix and post back. I've temporarily hacked a solution that's looking at job timestamps and deleting the lock file directly.

**Comment by manast** [comment]
@mfrye I am a bit concerned you are not using BullMQ optimally. For instance, stalled jobs do not "hang" a queue, they are just moved back to wait status. Let's assume that for some reason they are not moved to wait, then the queue should continue processing other jobs in the wait list. However, maybe your workers are not working properly, maybe they are not finishing or something. Also it bothers me that you use "job.waitUntilFinished()", in most scenarios, using that method is not the correct way to using the library.

**Comment by mfrye** [comment]
@manast I'd appreciate any advice on optimal way to use BullMQ, or if it's even a good tool for my use case.

I'm using BullMQ in a few different areas of my pipeline right now, but the main area I'm having issues is at the tail end. It's basically a scrape queue, where jobs can be added from internal processes (priority 3) or by a realtime request (priority 1). If it's a realtime request, `job.waitUntilFinished()` is used. Each job has 2 retries, and within a couple workers I use a separate rate limiter for requests to a few 3rd parties.

Inside the scrape queue, it was more or less working reliably prior to the last few weeks, though I've had interment issues at peak capacity. A couple weeks ago I added a new process that is a bit more CPU heavy, but did all the heavy work off the main thread using [workerpool](https://www.npmjs.com/package/workerpool). This worked fine locally, but the issues started after deploying to our ECS cluster.

I subsequently moved from workerpool to the sandbox pattern here, and I've been going through the the CPU heavy process to make it more async and avoid blocking the event loop as much as possible. No luck. 

I'm finding that:

- Jobs may or may not finish. Sometimes sandbox workers come back with a sigkill response.
- In each case, whether "complete" via erroring out or via sandbox kill, both seem to be stuck in active and aren't removed. 
- Once jobs are stuck, they are stuck. It appears they will sit there indefinitely before I manually intervene. 
- Restarting the worker instance seems to fix it and unblock stuck jobs.

**Comment by spearmootz** [comment]
I havent been able to see this happening anymore.   i did not change anything, i still have my code running checking to see if i have something in the queue but it doesnt say anything abnormal, it does show things being picked up as stalled and then it moves on. I am not looking at the environment as much so i am not sure this is an issue or not.

**Comment by manast** [comment]
@spearmootz in that case since there is no reproducible case for me to verify it I will close the issue. If anybody is facing something related please create a new issue with code that reproduces it.
